osdep/win32/smtc.cc

```diff
@@ -0,0 +1,542 @@
+            TranslateMessage(&msg);
+            DispatchMessageW(&msg);
+        }
+    } catch (const winrt::hresult_error& e) {
+        MP_ERR(&ctx, "%s: 0x%x - %ls\n", __func__, int32_t(e.code()), e.message().c_str());
+    }
+
+    ctx.close = true;
+    mpv_wakeup(mpv);
+    HWND hwnd = ctx.hwnd;
+    ctx.hwnd = nullptr;
+    if (ctx.thumb_async)
+        ctx.thumb_async.Cancel();
+    DestroyWindow(hwnd);
+    UnregisterClassW(wc.lpszClassName, HINSTANCE(&__ImageBase));
+
+    MP_THREAD_RETURN();
+}
+
+static MP_THREAD_VOID mpv_event_loop_fn(void *arg)
+{
+    mp_thread_set_name("smtc/mpv");
+    auto mpv = static_cast<mpv_handle *>(arg);
+    smtc_ctx ctx = {
+        .log = mp_client_get_log(mpv),
+        .mpv = mpv
+    };
+
+    // Create a dedicated window and event loop. We could use the mpv main window,
+    // but it is not always available, especially in audio-only/console mode.
+    mp_thread win_event_loop;
+    if (mp_thread_create(&win_event_loop, win_event_loop_fn, &ctx)) {
+        MP_ERR(&ctx, "Failed to create window event thread!\n");
+        goto error;
+    }
+
+    // It is recommended that you keep the system controls in sync with your
+    // media playback by updating these properties approximately every 5 seconds
+    // during playback and again whenever the state of playback changes, such as
+    // pausing or seeking to a new position.
+    // https://learn.microsoft.com/windows/uwp/audio-video-camera/system-media-transport-controls
+    // For simplicity we observe time-pos and duration as integers, so we get
+    // update every second, faster than recommended, but should be fine.
+
+    mpv_observe_property(mpv, 0, "current-tracks", MPV_FORMAT_NODE_MAP);
+    mpv_observe_property(mpv, 0, "duration", MPV_FORMAT_INT64);
+    mpv_observe_property(mpv, 0, "idle-active", MPV_FORMAT_FLAG);
+    mpv_observe_property(mpv, 0, "media-title", MPV_FORMAT_STRING);
+    mpv_observe_property(mpv, 0, "metadata", MPV_FORMAT_NODE_MAP);
+    mpv_observe_property(mpv, 0, "pause", MPV_FORMAT_FLAG);
+    mpv_observe_property(mpv, 0, "shuffle", MPV_FORMAT_DOUBLE);
+    mpv_observe_property(mpv, 0, "speed", MPV_FORMAT_DOUBLE);
+    mpv_observe_property(mpv, 0, "time-pos", MPV_FORMAT_INT64);
+    mpv_observe_property(mpv, 0, "track-list", MPV_FORMAT_NODE);
+    // TODO: Options are not observable, fix me!
+    mpv_observe_property(mpv, 0, "loop-file", MPV_FORMAT_DOUBLE);
+    mpv_observe_property(mpv, 0, "loop-playlist", MPV_FORMAT_DOUBLE);
+
+    while (!ctx.close) {
+        mpv_event *event = mpv_wait_event(mpv, -1);
+        if (event->event_id == MPV_EVENT_SHUTDOWN) {
+            if (ctx.hwnd)
+                PostMessageW(ctx.hwnd, WM_CLOSE, 0, 0);
+            break;
+        }
+        if (event->event_id == MPV_EVENT_PROPERTY_CHANGE ||
+            event->event_id == MPV_EVENT_PLAYBACK_RESTART)
+        {
+            if (ctx.hwnd)
+                SendMessageW(ctx.hwnd, WM_MP_EVENT, WPARAM(event), 0);
```

Comment:
> The convention for window messages is to use lparam for pointers. Using wparam looks weird to me.

I've been using both of them before, changed.

> TOCTOU 

I don't see how. This check only performs a check once the window is created on the other thread. The only way to initiate window destruction is from this thread, so no way it disappears while we are sending message. Unless destroyed externally, but this is not supported case.

I can add a mutex for initialization though, so we don't miss some events from mpv, but frankly, I doubt we manage to load file faster than this window is created, so I don't think it is really necessary.

Reply:
> I saw that the mpv event loop is ended after the WM_CLOSE message

Exactly, there is a break there. And yes, if something externally closes this window, thing would break, but I don't consider this a valid scenario. Else we would need to notify the other thread about it and in this case probably whole thing would need to be refactored to work.